server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.rank;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.VersionedNamedWriteable;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.search.SearchService;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * {@code RankContextBuilder} is used as a base class to manage input, parsing,
+ * and subsequent generation of appropriate contexts for handling searches that
+ * require multiple queries for global rank relevance.
+ *
+ * This class contains a {@code List<QueryBuilder>}s, size, and from members that
+ * sub-classes use to execute queries to generate ranking results. These are not part
+ * of ranking input, but instead generated from their respective values within a search
+ * query. We cannot generate the list of queries until after any kNN queries are executed.
+ */
+public abstract class RankBuilder<RCB extends RankBuilder<RCB>> implements VersionedNamedWriteable, ToXContentObject {
+
+    public static final ParseField WINDOW_SIZE_FIELD = new ParseField("window_size");
+
+    public static final int DEFAULT_WINDOW_SIZE = SearchService.DEFAULT_SIZE;
+
+    protected int windowSize = DEFAULT_WINDOW_SIZE;
```

Comment:
I made this private.

I switched to an ObjectParser because it's a better pattern for a Builder whereas ConstructingObjectParser is a better fit for an immutable object.

Reply:
I think it can be immutable if we move the values to the constructor. I'll figure out if there's any reason to not do this.

Reply:
`RankBuilder` and `RRFRankBuilder` now immutable.